### PR TITLE
Dcj/feature/js language detection

### PR DESCRIPTION
**Pull request by pfrincu-plenty**
### All changes meet the following requirements
- [ ] Changelog entry was added
- [ ] Changes have been documented
- [ ] Changes have been tested by the author
- [x] Changes have been tested by the reviewer

@plentymarkets/ceres-io

### CODE REVIEW

File: resources/views/Widgets/Header/LanguageDetectionWidget.twig [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    {{ Twig.do("set buttonTranslations = buttonTranslations | merge({(languageISO): trans('Ceres::Template.languageDetectionButton', [], languageISO ) })") }}
+{{ Twig.endfor() }}
+
+<div class="language-detection header-fw widget-{{ appearance }}{% if customClass | length > 0 %} {{ customClass }}{% endif %}">
+    <div class="container-max px-0 pr-lg-3">
+        {% if isPreview %}
+            <div class="align-center">
+                Language Detection
+            </div>
+        {% else %}
```
Comment (felixdausch) [inline resources/views/Widgets/Header/LanguageDetectionWidget.twig]: If possible, the preview should look like the production output of the widget. You should be able to use your Vue component in the preview as well. Maybe you need to check inside of your component if you are in the ShopBuilder to display the language picker even if the displayed language already matches the browser language. To do so, you can access `$ceres.isShopBuilder` inside of your Vue components.

File: resources/scss/ceres/widgets/Header/_language-detection-widget.scss [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      @include widget-bg-variant(rgba($info, .8));
+  }
+
+  &.widget-warning {
+      @include widget-bg-variant(rgba($warning, .8));
+  }
+
+  &.widget-danger {
+      @include widget-bg-variant(rgba($danger, .8));
+  }
```
Comment (felixdausch) [inline resources/scss/ceres/widgets/Header/_language-detection-widget.scss]: This isn't required. You can just add the class `bg-appearance` to apply the background color depending on the appearance-class

File: resources/js/src/app/components/pageDesign/LanguageDetection.vue [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  export default {
+    data () {
+      return {
+        languageText: '',
+        buttonText: '',
+        languageRedirect: '',
+        textLanguages: ''
+      }
+    },
+    props: ["redirect", "appearance", "texttranslations", "buttontranslations"],
```
Comment (felixdausch) [inline resources/js/src/app/components/pageDesign/LanguageDetection.vue]: Please use the object syntax to define the props of this component to have at least a type check on the prop values:

https://vuejs.org/v2/api/#props

File: resources/js/src/app/components/pageDesign/LanguageDetection.vue [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+          }
+        }
+      },
+      deactivateRedirection() {
+        this.languageText = null;
+        window.localStorage.setItem('redirectDeactivated', true);
+      },
+      setAppearance() {
+        this.appearance = this.$props.appearance;
+      },
```
Comment (felixdausch) [inline resources/js/src/app/components/pageDesign/LanguageDetection.vue]: This shouldn't be necessary. You can access your props in the same way like you can access the components data. So `{{ appearance }}` will work even without this function

File: resources/js/src/app/components/pageDesign/LanguageDetection.vue [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      setText(lang) {
+        const textLangObj = JSON.parse(JSON.stringify(this.$props.texttranslations));
```
Comment (felixdausch) [inline resources/js/src/app/components/pageDesign/LanguageDetection.vue]: Why the json encoding/decoding here? The prop-value should already contain an accessible javascript object.

File: resources/js/src/app/components/pageDesign/LanguageDetection.vue [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        this.languageText = textLangObj[lang];
+      },
+      setButton(lang) {
+        const buttonLangObj = JSON.parse(JSON.stringify(this.$props.buttontranslations));
```
Comment (felixdausch) [inline resources/js/src/app/components/pageDesign/LanguageDetection.vue]: Same as in `setText()`

File: resources/js/src/app/components/pageDesign/LanguageDetection.vue [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+        for (let i = 0; i < window.navigator.languages.length; i++) {
+          const langObject= window.navigator.languages[i].split('-')
+          const linkTag = document.querySelector('link[hreflang="' + langObject[0] +'"]')
+
+          if (App.language !== langObject[0] && linkTag) {
+            if (this.$props.redirect){
+              if (!window.localStorage.getItem('redirectActive')){
+                window.localStorage.setItem('redirectActive', true);
+                window.location.href = linkTag.getAttribute('href')
```
Comment (felixdausch) [inline resources/js/src/app/components/pageDesign/LanguageDetection.vue]: Please use `navigateTo()` from UrlService.js

File: resources/js/src/app/components/pageDesign/LanguageDetection.vue [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+        this.setAppearance();
+
+        for (let i = 0; i < window.navigator.languages.length; i++) {
+          const langObject= window.navigator.languages[i].split('-')
+          const linkTag = document.querySelector('link[hreflang="' + langObject[0] +'"]')
+
+          if (App.language !== langObject[0] && linkTag) {
+            if (this.$props.redirect){
+              if (!window.localStorage.getItem('redirectActive')){
```
Comment (felixdausch) [inline resources/js/src/app/components/pageDesign/LanguageDetection.vue]: Why is this entry in the local storage required?

**Review by felixdausch — DISMISSED** [review]
Please move the different cases during initialization to different methods to keep the code more readable.
Do you really need two flags to be stored in the localStorage? If yes, please find more speaking names that describe better, whats the purpose of each of these flags.

Finally we need to take care of the GDPR. So the flags are only allowed to be stored if the user has accepted the usage. For this you need to hook in the cookie bar of Ceres as described here:
https://developers.plentymarkets.com/en-gb/developers/main/plentyshop-plugins/how-to-register-cookies.html

File: src/Widgets/Header/LanguageDetectionWidget.php [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        return WidgetDataFactory::make("Ceres::LanguageDetectionWidget")
+            ->withLabel("Widget.languageDetectionWidgetLabel")
+            /*
+             * TODO: replace the image with a proper one
+             */
+            ->withPreviewImageUrl("/images/widgets/top-bar.svg")
+            ->withType(WidgetTypes::HEADER)
+            ->withCategory(WidgetCategories::HEADER)
+            ->withPosition(0)
```
Comment (cjohannsen) [inline src/Widgets/Header/LanguageDetectionWidget.php]: this widget should not be the first one in the list (position 0)

File: resources/js/src/app/components/pageDesign/LanguageDetection.vue [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    props: ["redirect", "appearance", "texttranslations", "buttontranslations"],
+    mounted() {
+      this.initializeComponent()
+    },
+    methods: {
+      initializeComponent() {
+
+        this.setAppearance();
+
+        for (let i = 0; i < window.navigator.languages.length; i++) {
```
Comment (cjohannsen) [inline resources/js/src/app/components/pageDesign/LanguageDetection.vue]: do not use for loops if not necessary, please use window.navigator.languages.forEach(language => ...)

**Review by cjohannsen — CHANGES_REQUESTED** [review]

File: resources/scss/ceres/widgets/Header/_language-detection-widget.scss (hunk 2 above)
Comment (felixdausch) [reply resources/scss/ceres/widgets/Header/_language-detection-widget.scss]: @Tim-M-S can this be removed regarding to your changes on the widget template?

File: resources/js/src/app/components/pageDesign/LanguageDetection.vue (hunk 7 above)
Comment (felixdausch) [reply resources/js/src/app/components/pageDesign/LanguageDetection.vue]: See https://github.com/plentymarkets/plugin-ceres/blob/stable/resources/js/src/app/services/UrlService.js#L162

File: resources/lang/de/Widget.properties [hunk 11]
```diff
@@ -1012,3 +1012,7 @@ preloadImageLabel = "Bild vorladen"
 preloadImageTooltip = "Das Bild wird beim Laden der Seite priorisiert. Diese Option sollte nur aktiviert werden, wenn sich das Widget in dem Bereich der Seite befindet, der beim Aufruf der Seite initial sichtbar ist."
 maxPreloadLabel = "Anzahl der vorgeladenen Bilder"
 maxPreloadTooltip = "Anzahl der vorgeladenen Bilder"
+; LanguageDetection
+languageDetectionLabel = "Spracherkennung"
+languageDetectionWidgetLabel = "Spracherkennung"
+languageDetectionRedirect = "Automatische Weiterleitung aktivieren"
```
Comment (Tim-M-S) [inline resources/lang/de/Widget.properties]:
```suggestion
languageDetectionRedirect = "Automatische Weiterleitung aktivieren"

```

**Review by Tim-M-S — DISMISSED** [review]

**Review by cjohannsen — DISMISSED** [review]

**Review by Tim-M-S — APPROVED** [review]

**Pull request merged**